idn/punycode.go

```diff
@@ -30,6 +31,12 @@ const (
 // This function would return an empty string result for domain names with
 // invalid unicode strings. This function expects domain names in lowercase.
 func ToPunycode(s string) string {
+	// Early check to see if encoding is needed.
+	// This will prevent making heap allocations when not needed.
+	if !needpuny(s) {
+		return strings.ToLower(s)
```

Comment:
I would suggest removing the ToLower and documenting that upper case characters might be lowercased in the result.